packages/gatsby/src/internal-plugins/query-runner/query-watcher.js

```diff
@@ -132,22 +101,42 @@ const handleQuery = (
 const updateStateAndRunQueries = isFirstRun => {
   const snapshot = getQueriesSnapshot()
   return queryCompiler().then(queries => {
+    // If there's an error while extracting queries, the queryCompiler returns false
+    // or zero results.
+    // Yeah, should probably be an error but don't feel like threading the error
+    // all the way here.
+    if (!queries || queries.size === 0) {
+      return null
+    }
     handleComponentsWithRemovedQueries(snapshot, queries)
 
+    // Run action for each component
+    const { components } = snapshot
+    components.forEach(c =>
+      boundActionCreators.queryExtracted({
+        componentPath: c.componentPath,
+        query: queries.get(c.componentPath)?.text || ``,
+      })
+    )
+
     let queriesWillNotRun = false
     queries.forEach((query, component) => {
       const queryWillRun = handleQuery(snapshot, query, component)
 
       if (queryWillRun) {
         watchComponent(component)
-      } else if (isFirstRun) {
+        // Check if this is a page component.
+        // If it is and this is our first run during bootstrap,
+        // show a warning about having a query in a non-page component.
+      } else if (isFirstRun && !snapshot.components.has(component)) {
         report.warn(
           `The GraphQL query in the non-page component "${component}" will not be run.`
         )
         queriesWillNotRun = true
       }
     })
 
+    // TODO restore this logic for page components.
```

Comment:
will this cause a regression? (should we address it in this PR?)

Reply:
Oh actually already did this.